### PR TITLE
[c10d] Adding setTimeout option in Store

### PR DESCRIPTION
**Pull request by teng-li**
This will allow users to set customized timeout option for the store.

Tested by my own debug print to make sure that C++ actually used the timeout

### CODE REVIEW

File: torch/lib/c10d/FileStore.cpp [hunk 1]
```diff
@@ -199,13 +199,19 @@ void FileStore::set(const std::string& key, const std::vector<uint8_t>& value) {
...
     File file(path_, O_RDONLY);
     auto lock = file.lockShared();
     auto size = file.size();
     if (size == pos_) {
       // No new entries; release the shared lock and sleep for a bit
       lock.unlock();
+      const auto elapsed = std::chrono::duration_cast<std::chrono::seconds>(
+          std::chrono::steady_clock::now() - start);
+      if (timeout_ != kNoTimeout && elapsed > timeout_) {
+        throw std::runtime_error("Get timeout");
```
Comment (pietern) [inline torch/lib/c10d/FileStore.cpp]: `"Timeout waiting for key: " + key`

**Review by pietern — COMMENTED** [review]
Perhaps also update the error message in the TCPStore if it doesn't include the key name in the exception.

File: torch/lib/c10d/FileStore.cpp (hunk 1 above)
Comment (teng-li) [reply torch/lib/c10d/FileStore.cpp]: Addressed

**Comment by teng-li** [comment]
TCPStore doesn't have the similar exception message, I will import and update internal code

**Pull request closed without merging**